### PR TITLE
TL/SHARP: sharp coll context/tree  per team

### PR DESCRIPTION
**Pull request by bureddy**
Option to use sharp coll context(tree/job) per team ( similar NCCL SHARP model)

### CODE REVIEW

File: src/components/tl/sharp/tl_sharp.c [hunk 1]
```diff
@@ -48,6 +48,12 @@ static ucc_config_field_t ucc_tl_sharp_context_config_table[] = {
      ucc_offsetof(ucc_tl_sharp_context_config_t, uprogress_num_polls),
      UCC_CONFIG_TYPE_UINT},
 
+    {"CONTEXT_PER_TEAM", "n",
+     "Number of polls to do before calling user progress",
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp.c:52]: Incorrect parameter description

File: src/components/tl/sharp/tl_sharp_coll.c [hunk 2]
```diff
@@ -82,21 +82,32 @@ static inline ucc_status_t ucc_tl_sharp_status_to_ucc(int status)
...
 
     if (length < ctx->cfg.reg_threshold) {
         *reg = &ucc_tl_sharp_reg_null;
         return UCC_OK;
     }
 
-    if (ctx->rcache) {
-        status = ucc_rcache_get(ctx->rcache, (void *)addr, length, NULL,
+    if (ctx->sharp_context  != NULL) {
+        sharp_ctx = ctx->sharp_context;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_coll.c:102]: maybe always set team->sharp_ctx during team creation. If CTX_PER_TEAM is not set then the ptr will point to global context. Then you don't need these branches, and the ucc_tl_sharp_mem_register can just take team

File: src/components/tl/sharp/tl_sharp_context.c [hunk 3]
```diff
@@ -275,20 +261,40 @@ static ucc_rcache_ops_t ucc_tl_sharp_rcache_ops = {
...
     struct sharp_coll_init_spec  init_spec = {0};
     ucc_tl_sharp_lib_t          *lib       = ucc_derived_of(sharp_ctx->super.super.lib,
                                                             ucc_tl_sharp_lib_t);
     ucc_status_t status;
 
     init_spec.progress_func                  = NULL;
-    init_spec.world_rank                     = UCC_TL_CTX_OOB(sharp_ctx).oob_ep;
     init_spec.world_local_rank               = 0;
-    init_spec.world_size                     = UCC_TL_CTX_OOB(sharp_ctx).n_oob_eps;
     init_spec.group_channel_idx              = 0;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_context.c:295]: duplicate of group_channel_idx?

File: src/components/tl/sharp/tl_sharp_context.c [hunk 4]
```diff
@@ -375,49 +381,51 @@ UCC_CLASS_INIT_FUNC(ucc_tl_sharp_context_t,
...
+    if (lib->cfg.use_internal_oob) {
+        sharp_ctx->oob_ctx.subset.map.ep_num =
+            UCC_TL_CTX_OOB(sharp_ctx).n_oob_eps;
+        sharp_ctx->oob_ctx.subset.map.type   = UCC_EP_MAP_FULL;
+        sharp_ctx->oob_ctx.subset.myrank     = UCC_TL_CTX_OOB(sharp_ctx).oob_ep;
+    } else {
+        sharp_ctx->oob_ctx.oob      = &UCC_TL_CTX_OOB(sharp_ctx);
+    }
+
+    if (sharp_ctx->cfg.context_per_team) {
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_context.c:407]: why not making this check before line 398?

File: src/components/tl/sharp/tl_sharp_team.c [hunk 5]
```diff
@@ -15,7 +15,9 @@ UCC_CLASS_INIT_FUNC(ucc_tl_sharp_team_t, ucc_base_context_t *tl_context,
     ucc_tl_sharp_context_t         *ctx =
         ucc_derived_of(tl_context, ucc_tl_sharp_context_t);
     struct sharp_coll_comm_init_spec comm_spec;
-    int                              ret;
+    int                             ret;
+    ucc_status_t                    status;
+    struct sharp_coll_context      *sharp_ctx = ctx->sharp_context;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_team.c:20]: initialized vars go first

**Review by vspetrov — CHANGES_REQUESTED** [review]
Copyrights 2022

File: src/components/tl/sharp/tl_sharp_coll.c [hunk 6]
```diff
@@ -82,21 +82,32 @@ static inline ucc_status_t ucc_tl_sharp_status_to_ucc(int status)
...
+        rcache = ctx->rcache;
+    } else {
+        sharp_ctx = team->sharp_context;
+        rcache = team->rcache;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_coll.c:106]: align =

File: src/components/tl/sharp/tl_sharp_coll.c [hunk 7]
```diff
@@ -111,28 +122,38 @@ ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
...
     return UCC_OK;
 }
 
 static ucc_status_t
-ucc_tl_sharp_mem_deregister(ucc_tl_sharp_context_t *ctx,
+ucc_tl_sharp_mem_deregister(ucc_tl_sharp_context_t *ctx, ucc_tl_sharp_team_t *team,
                             ucc_tl_sharp_reg_t *reg)
 {
     ucc_tl_sharp_rcache_region_t *region;
+    ucc_rcache_t *rcache;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_coll.c:137]: align local vars declarations

File: src/components/tl/sharp/tl_sharp_coll.c [hunk 8]
```diff
@@ -111,28 +122,38 @@ ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
...
+    struct sharp_coll_context *sharp_ctx;
 
     if (reg == &ucc_tl_sharp_reg_null) {
         return UCC_OK;
     }
 
-    if (ctx->rcache) {
+    if (ctx->sharp_context  != NULL) {
+        sharp_ctx = ctx->sharp_context;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_coll.c:145]: align =

File: src/components/tl/sharp/tl_sharp_team.c [hunk 9]
```diff
@@ -24,6 +26,8 @@ UCC_CLASS_INIT_FUNC(ucc_tl_sharp_team_t, ucc_base_context_t *tl_context,
 
     UCC_CLASS_CALL_SUPER_INIT(ucc_tl_team_t, &ctx->super, params);
 
+    self->sharp_context = NULL;
```
Comment (vspetrov) [inline src/components/tl/sharp/tl_sharp_team.c:29]: align =

**Comment by bureddy** [comment]
bot:retest

**Review by vspetrov — APPROVED** [review]

**Review by manjugv — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**